### ISSUE TITLE
sql: can OOM a cluster with concurrent TPC-H load

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#64906]
It's easy to overwhelm a cluster by causing it to OOM when throwing highly concurrent TPC-H load at the cluster.

Instructions to repro:

1. start roachprod cluster
2. `./cockroach workload init tpch --data-loader=import`
3. `./cockroach workload run tpch --queries 1 --concurrency=1024 --tolerate-errors`
4. wait. eventually, some of the nodes will probably be oomkilled.

--- 

I spent a good deal of time today playing with OOM conditions in Cockroach in general, after learning about this fun issue: https://github.com/golang/go/issues/42805. This issue talks about how spiky allocations can cause issues in Go because of the pacer implementation. I think we see this in practice all the time. Someone wrote a program that is a "watchdog" which runs GC manually when certain watermarks are exceeded (https://github.com/raulk/go-watchdog) so I thought I'd play with this idea.

Things I learned:

1. If you add a goroutine that calls `runtime.GC()` once every second the cluster does not do much better, although it lasts a little bit longer
2. The accounted-for SQL memory stays relatively constant
3. Turning off the load allows the memory usage to return to baseline relatively quickly
4. The GC allows the nodes to return to ~50% memory usage for a while, but it slowly begins inching up from that 50% baseline until it eventually dies. I think the 50% is expected (25% for SQL, 25% for pebble)
5. The cluster can do zero work this entire time (all queries either return SQL memory reservation errors or hang)
6. heap profiles show "expected size" profiles, with all memory in expected places
7. the major unaccounted part of the profiles is in the scan buffer from pebble

When the oomkiller strikes, it can be very sudden. One time I saw the memory go from less than 50% heap usage to dead without another sample (was using https://github.com/arl/statsviz which is an amazing project that we should integrate into Cockroach's DB Console, which samples 1/sec when you have it open in the browser). This was pretty scary - nothing that I know of allocates so quickly. What could it mean?

In general I think we have 2 avenues we should poke at:

1. what is that huge allocation that sometimes strikes right before oomkill? It could very well be a red herring, but I'm not sure.

![image](https://user-images.githubusercontent.com/43821/117521164-b9a2de00-af82-11eb-85bb-3a0d8eb9f89c.png)

2. should we account for pebble scan batches? it's odd because we immediately transfer all of that memory into an accounted-for SQL batch, and then allow the pebble scan batch to be freed, but perhaps the moments at which both the SQL batch and the pebble scan batch are live are too long when the cluster is overloaded?
![image](https://user-images.githubusercontent.com/43821/117521150-b0b20c80-af82-11eb-80ef-ee4389a34c49.png)

Jira issue: CRDB-7339

**Comment by jordanlewis** [comment]
Related issue is #64916. After fixing that issue, I observe better memory behavior, but it is not completely better - I can still crash the cluster with high concurrency on certain TPCH queries (I've been playing with query 18), but it takes somewhat longer now.

---- 

I noticed something interesting: there might be a cascading failure behavior here, caused by the differences in performance when distsql plans are correct (table readers on leaseholders) vs when they are incorrect (table readers are not on leaseholders).

DistSQL will plan table readers on leaseholders at plan time. But, if the cluster starts getting overwhelmed and leaseholders move around due to liveness issues, then by the time a plan is run, the leaseholder may have changed. DistSQL does not know how to dynamically replan a plan: it is going to run the same plan that it started with, even if leaseholders change.

The performance difference when a table reader is on the leaseholder and when it's not is quite significant. This is because, when the table reader is not on the leaseholder, it has to do a gRPC request to the leaseholder to fetch data.

Also, when table readers are on leaseholders, there is 1 memory buffer unaccounted for: the pebble one discussed above.

On the other hand, when table readers are *not* on leaseholders, there are *2* memory buffers unaccounted for: 
1. the gRPC buffers that are used to receive batches from the leaseholder (always allocated and thrown away in `recv`, shoutout to this classic issue https://github.com/grpc/grpc-go/issues/1455)
2. the `BatchResponse` protobuf that is decoded into from step 1 (see `BatchResponse.Unmarshal` generated code)

This is twice the unaccounted-for memory!

So I'm theorizing that, as the performance of a cluster begins to falter due to lack of resources, leaseholders will change around, which causes worse memory behavior (not to mention slower reads in the first place due to extra serialization and network IO caused by the necessity for the remote hop), which overwhelms the cluster further, until it eventually blows up due to things being slow and slop in the memory accounting getting worse as the plans aren't accurate with regards to leaseholder info.

**Comment by jordanlewis** [comment]
Example of the "double unaccounted-for memory" as it would appear in a pprof flamegraph:

![image](https://user-images.githubusercontent.com/43821/117550536-53b46600-b017-11eb-9b86-6b72dda9dd76.png)

When a tablereader is planned on the leaseholder, neither of these stacks will appear: instead, you'll see the `pebbleResults.put` stack like in the flamegraph higher up in the thread.

**Comment by petermattis** [comment]
Are you aware of [`GODEBUG=allocfreetrace=1`](https://golang.org/pkg/runtime/)? It may be infeasible to use this facility due to the volume of output, but it is another tool for tracing down a hard to discern sudden spike in memory usage. If the volume of output is too large, you could hack the Go runtime to only output allocations/free larger than some threshold size.

**Comment by jordanlewis** [comment]
Yes, I was considering exactly this strategy yesterday, but I haven't tried it yet.

At the moment, I am theorizing that the large, sudden allocations are caused by gRPC. gRPC has zero memory pressure flow control, which means that a node that sends several batch requests to another node, which is stuck, but suddenly gets unstuck, will happily eat itself to death when all of the batch responses start actually coming back.

**Comment by joshimhoff** [comment]
With the two concrete issues mentioned (pebble scan unaccounted for mem usage & table reader != leaseholder leading to grpc buffer + `BatchResponse` unaccounted for mem usage), will admission control keep the lack of accounting from leading to OOMs, at least maybe?

1. pebble scan unaccounted for mem usage 

Obviously the accounting being missing could / should (well, y'all tell me) be corrected. But for this to cause OOMs, it sounds like you need a very high level of concurrency. Perhaps with admission control, you will start queuing work before reaching that level of concurrency. Perhaps also, CPU overload will be kept low enough that the below theory will not happen:

> perhaps the moments at which both the SQL batch and the pebble scan batch are live are too long when the cluster is overloaded?

2. table reader != leaseholder leading to grpc buffer + `BatchResponse` unaccounted for mem usage

Again the accounting being missing could / should be corrected. But IIUC admission control will queue work so as to reserve enough resources to maintain liveness. So leaseholders won't move around, at least not for the concrete reason of failing to maintain liveness.

Thoughts? Maybe there are more "OOM when highly overloaded due to DML only workload" bugs that will also be "fixed" by admission control, as there will be a limit to how overloaded CRDB ever gets due to queueing at the application level.

**Comment by jordanlewis** [comment]
> But for this to cause OOMs, it sounds like you need a very high level of concurrency. Perhaps with admission control, you will start queuing work before reaching that level of concurrency.

I agree that admission control will likely help us get out of this situation. But I'm not sure if it would be enough on its own, depending on how it deals with long-running, giant distsql requests, which are pretty different than kv-style load.

----

I think we probably need to hook gRPC up to our memory accounting. An appropriate place would be here. As you can see, we read a message length off of the wire, then unconditionally allocate that size of message.

https://github.com/grpc/grpc-go/blob/d2d6bdae07e844b8a3502dcaf00dc7b1b5519a59/rpc_util.go#L562-L575

I'm less convinced that we need to hook pebble up to our memory accounting just yet. I think that the memory accounting delay here is not as significant as that of gRPC, but more experimentation is needed.

**Comment by adityamaru** [comment]
@jordanlewis I went down a similar path when debugging an index backfiller OOM a few months ago - https://github.com/cockroachdb/cockroach/issues/58933#issuecomment-788065865. The only thing growing unbounded was the pebble scan batch. We _fixed_ it by using a new Fetcher per index backfill batch, rather than a single, long-lived one - https://github.com/cockroachdb/cockroach/pull/61366, but I couldn't understand why the underlying pebble batch wasn't being freed. Not much to add to your investigation but I'm keenly following along in case my fix just papered over something more sinister.

**Comment by jordanlewis** [comment]
Oh, thanks for the tip! That does seem extremely suspicious. I see a couple of places where the fetcher could be holding onto memory from a pebble batch that we could be clearing, but doing that wouldn't hold more than 1 extra pebble batch in memory at a time, so I'm not totally sure how your fix could have solved a problem that was more than accidental 2x memory - I don't see how we could have unbounded growth.

I'm thinking about the `indexKey`, `kv`, `keyRemainingBytes` fields in particular - those are slices into underlying memory that isn't copied out of the pebble batch for performance reasons. We should clear those when the fetcher is finished and that will save us some memory, possibly up to 2x.

I think the cfetcher has similar issues, in the `nextKV` and `lastRowPrefix` fields.

Also, there are some similar issues in `kv_batch_fetcher`, which stores its `responses` and `remainingBatches` slices in a queue pattern, and continually slices off the front of them to do the queuing. But, it doesn't `nil` out the front elements of the queue. I actually made a change to fix this earlier on the branch that I've been playing with for this issue, but it didn't improve anything. It's possible that this was the root cause of #58933, though - it could look like unbounded growth depending on how Go's slice reallocation algorithm works when you need more space in a slice that you've also been "cutting off the front of".

Diff for this last thing that I've been playing with:

```
diff --git a/pkg/sql/row/kv_batch_fetcher.go b/pkg/sql/row/kv_batch_fetcher.go
index 6197c2d746..5a4e5a95b3 100644
--- a/pkg/sql/row/kv_batch_fetcher.go
+++ b/pkg/sql/row/kv_batch_fetcher.go
@@ -504,11 +506,13 @@ func (f *txnKVFetcher) nextBatch(
 ) (ok bool, kvs []roachpb.KeyValue, batchResponse []byte, origSpan roachpb.Span, err error) {
        if len(f.remainingBatches) > 0 {
                batch := f.remainingBatches[0]
+               f.remainingBatches[0] = nil
                f.remainingBatches = f.remainingBatches[1:]
                return true, nil, batch, f.origSpan, nil
        }
        for len(f.responses) > 0 {
                reply := f.responses[0].GetInner()
+               f.responses[0] = roachpb.ResponseUnion{}
                f.responses = f.responses[1:]
                origSpan := f.requestSpans[0]
                f.requestSpans = f.requestSpans[1:]
@@ -517,12 +521,14 @@ func (f *txnKVFetcher) nextBatch(
                case *roachpb.ScanResponse:
                        if len(t.BatchResponses) > 0 {
                                batchResp = t.BatchResponses[0]
+                               t.BatchResponses[0] = nil
                                f.remainingBatches = t.BatchResponses[1:]
                        }
                        return true, t.Rows, batchResp, origSpan, nil
                case *roachpb.ReverseScanResponse:
                        if len(t.BatchResponses) > 0 {
                                batchResp = t.BatchResponses[0]
+                               t.BatchResponses[0] = nil
                                f.remainingBatches = t.BatchResponses[1:]
                        }
                        return true, t.Rows, batchResp, origSpan, nil
```

**Comment by jordanlewis** [comment]
It's completely sad that Go doesn't have any tools that allow you to explore a heap dump and understand what is retaining different objects. A problem like this would take, seriously, all of 15 minutes to solve with Java and YourKit. </rant>

**Comment by jordanlewis** [comment]
Btw, I filed https://github.com/golang/go/issues/45984 recently because the Go `viewcore` program doesn't work on the cores we emit for whatever reason.

**Comment by joshimhoff** [comment]
> It's completely sad that Go doesn't have any tools that allow you to explore a heap dump and understand what is retaining different objects.

Was reading about this too. Sad!

> I agree that admission control will likely help us get out of this situation. 

One concrete Q I have is: When you are hitting these OOMs, what CPU & disk IOPS utilization are ya at? High enough that we would be queueing work in an admission control world.

> But I'm not sure if it would be enough on its own, depending on how it deals with long-running, giant distsql requests, which are pretty different than kv-style load.

Can you say more about this? Do you have a concrete worry about such load or is it more uncertainty about how admission control (at the SQL level) will interact with such load?

**Comment by jordanlewis** [comment]
> One concrete Q I have is: When you are hitting these OOMs, what CPU & disk IOPS utilization are ya at? High enough that we would be queueing work in an admission control world?

I'm not sure but I will check. I'm almost sure we'd be queueing work here. I didn't make it clear, but sending 1024 concurrent TPCH query 18s at a 3 node cluster is a completely absurd thing to do be doing. It's full overload - there is no chance for the cluster to be able to respond to these queries successfully without queueing. The absurdity of this workload is why it's a good testbed for these OOMs, but also probably (hopefully) why admission control would help protect us here in the future.

> Can you say more about this? Do you have a concrete worry about such load or is it more uncertainty about how admission control (at the SQL level) will interact with such load?

It's uncertainty about where the admission control spigots will be placed. Since these queries are large and relatively long-running, and since we kick them off all at once from an idle system, I don't think admission control would necessarily notice to reject or queue anything at first. So I'm uncertain whether anybody will detect the overload and be able to error out some of these queries while they're running.

**Comment by jordanlewis** [comment]
#65881 improves the behavior for query 18.

Query 4 can still OOM a server in seconds. I think it's because lookup join does not play nicely and is not well-memory-accounted. I have filed a new issue:

#65904

**Comment by jordanlewis** [comment]
I've re-tested query 4 after @rytaft merged #67771 to fix #65904.

The situation is somewhat better! But not altogether solved. The server no longer instantly OOMs with high query 4 concurrency - instead, it OOMs after several minutes.

I noticed that, at concurrency 128 with the default roachprod machine sizes, the system reaches somewhat of a steady state - there's high memory usage, but the garbage collector can keep up, and there are no query rejections due to memory accounting.

I analyzed a core dump mid-run, and found that there was about 3.4 GB live memory, and 3.4 GB garbage:

```
(viewcore) breakdown
 all                   13 GB 100.00%
   text                96 MB   0.73%
   readonly            61 MB   0.47%
   data                13 MB   0.10%
   bss                1.1 GB   8.56% (grab bag, includes OS thread stacks, ...)
   heap                11 GB  87.33%
     in use spans     7.0 GB  53.28%
       alloc          6.9 GB  52.44%
         live         3.4 GB  26.18%
         garbage      3.4 GB  26.26%
       free            71 MB   0.54%
       round           39 MB   0.30%
     manual spans      24 MB   0.18% (Go stacks)
       alloc           23 MB   0.18%
       free           553 kB   0.00%
     free spans       4.4 GB  33.87%
       retained       4.1 GB  31.67% (kept for reuse by Go)
       released       288 MB   2.21% (given back to the OS)
   ptr bitmap         356 MB   2.73%
   span table          11 MB   0.09%
```

However, according to the SQL memory graph that tracks the root memory account, we should only be using 2GB of memory.

Here's a `viewcore` histogram of the objects that make up 80% of the heap. Notice that nearly 25% (800MB) of the live heap is made up of `roachpb.Span`.

```
(viewcore) histo --top=100
   count   size  bytes live%  sum% type
     182 2.5 MB 459 MB 13.43 13.43 [1+52564?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
     128 2.8 MB 353 MB 10.33 23.76 [1+57513?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
  419366  768 B 322 MB  9.42 33.18 [16]github.com/cockroachdb/cockroach/pkg/sql/rowenc.EncDatum
    3515  66 kB 230 MB  6.74 39.92 [65536]uint8
     128 1.5 MB 194 MB  5.67 45.59 [63146]github.com/cockroachdb/cockroach/pkg/sql/rowenc.EncDatumRow
   64604 2.3 kB 149 MB  4.35 49.94 [1+47?]github.com/cockroachdb/cockroach/pkg/sql/rowenc.EncDatum
      31 3.8 MB 117 MB  3.42 53.37 unk3776512
 7161156   16 B 115 MB  3.35 56.72 unk16
      11 7.0 MB  77 MB  2.24 58.96 unk6971392
  573825  128 B  73 MB  2.15 61.11 unk128
 1143465   64 B  73 MB  2.14 63.25 unk64
       2  34 MB  67 MB  1.96 65.21 [33554432]uint8
      90 737 kB  66 MB  1.94 67.15 [2194]bucket<string,[]int>
 1315978   48 B  63 MB  1.85 69.00 unk48
  558233  112 B  62 MB  1.83 70.83 unk112
    6957 8.2 kB  57 MB  1.67 72.49 [1+1023?]github.com/cockroachdb/cockroach/pkg/sql/sem/tree.DInt
      97 565 kB  55 MB  1.60 74.10 [1+11775?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
  532090   96 B  51 MB  1.49 75.59 unk96
      86 541 kB  46 MB  1.36 76.95 [1+11263?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
    2796  16 kB  46 MB  1.34 78.29 unk16384
  266561  144 B  38 MB  1.12 79.41 unk144
 2393405   16 B  38 MB  1.12 80.53 [1+15?]uint8
```

Using `peek` in viewcore, I determined that the spans are half retained by `txnKVFetcher`, and half by `defaultSpanGenerator`. @rytaft's recent patch accounts for the memory in `defaultSpanGenerator`, but any spans in `txnKVFetcher` are *not* accounted for:

```
(viewcore) peek [1+52564?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
 count   size  bytes  live%   sum% type
   182 2.5 MB 459 MB 100.00 100.00 github.com/cockroachdb/cockroach/pkg/sql/row.txnKVFetcher
Total: 459 MB
 count   size  bytes  live%   sum% type
   182 2.5 MB 459 MB 100.00 100.00 [1+52564?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
Total: 459 MB
   count size  bytes live%   sum% type
 7180092 16 B 115 MB 96.24  96.24 unk16
  280280 16 B 4.5 MB  3.76  99.99 [16]uint8
     407 16 B 6.5 kB  0.01 100.00 [1+15?]uint8
       9 16 B  144 B  0.00 100.00 [6+10?]uint8
Total: 119 MB
(viewcore) peek [1+57513?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
 count   size  bytes  live%   sum% type
   128 2.8 MB 353 MB 100.00 100.00 github.com/cockroachdb/cockroach/pkg/sql/rowexec.defaultSpanGenerator
Total: 353 MB
 count   size  bytes  live%   sum% type
   128 2.8 MB 353 MB 100.00 100.00 [1+57513?]github.com/cockroachdb/cockroach/pkg/roachpb.Span
Total: 353 MB
   count size  bytes live%   sum% type
 5990848 16 B  96 MB 89.27  89.27 unk16
  624502 16 B  10 MB  9.31  98.58 [16]uint8
   95553 16 B 1.5 MB  1.42 100.00 [1+15?]uint8
       9 16 B  144 B  0.00 100.00 [6+10?]uint8
Total: 107 MB
```

I noticed a few things from here:

1. `joinReader` isn't freeing its `spanGenerators` when it's finished with its work, so similar to #66106, once the `joinReader` is finished, there will be some time when it's retaining a bunch of memory that isn't accounted for.
2. `txnKVFetcher` is retaining 2x the necessary span memory, and could be freeing what it needs to retain more quickly.

After attempting some fixes for this (#68044), here's the memory graph for 128 concurrency:

![image](https://user-images.githubusercontent.com/43821/126926998-c6285e15-feb8-4cb1-a324-eb6ae001b614.png)

At 1024 concurrency, nothing much changes - most queries give back "no inbound stream connection", inbox errors, or memory accounting errors for several minutes, until the server crashes:

![image](https://user-images.githubusercontent.com/43821/126927304-0d501130-7416-4ba2-a0fd-9039092bfa99.png)

Doing another heap analysis at this point gave confusing results - the histogram showed a bunch of unknown/garbage types, which I think mostly mean that they're stack objects, but not sure. Will continue this another time.

**Comment by yuzefovich** [comment]
I think I've made major progress on this issue this week.

When I cherry-picked #70608, #70622, and #70729 onto the release-21.2 branch, the 3 node roachprod cluster with the default spec is now able to handle TPCH with concurrency 256 for two hours without a single crash (occasionally it does come quite close to the OOM scenario, but GC is able to keep up). All connections start out executing queries in the same fashion (i.e. Q1, Q2, ...), but eventually they will become mis-aligned, so it'll be a mix of queries running at the same time. The CPU is pegged at 100% on all nodes, while RAM is hovering in 10-13.5GB range (the machines have about 14GB of available).

![Screen Shot 2021-09-24 at 6 47 30 PM](https://user-images.githubusercontent.com/5135385/134753919-4f83e959-a59d-4633-bf02-b40820d91acf.png) 

[Here is the output of the workload](https://gist.githubusercontent.com/yuzefovich/46b75158b7500f1ca7469d816b0d80c7/raw/22adf1f9528c6842bf3e32efb939aa34eea7b32a/tpch%2520concurrency) which shows that the cluster is able to serve some queries correctly while hitting the memory budget errors on others (ignore the query 15 which creates and drops a view). The workload command I used was `./workload run tpch $(cat pgurls) --tolerate-errors --ramp=2m --duration=3h --concurrency=256`.

I imagine that #70648 could help a bit too (to spread the load more evenly.

**Comment by rytaft** [comment]
Great news! Exciting to see this work bearing fruit.

**Comment by yuzefovich** [comment]
Over in #98428 I observed that we can now sustain `tpch_concurrency` roachtest with concurrency of 1000. It is still possible to OOM cockroach with it, but GOMEMLIMIT (introduced in #97666) saves us vast majority of the time. This OOM happens when the nodes are so overloaded that extremely rare bugs show up, so I think we can call this issue solved. 🎉 

@jordanlewis thoughts on closing this?

**Comment by jordanlewis** [comment]
Amazing work, and great news!

I'm fine to see this get closed if you're feeling satisfied, go ahead.

**Comment by yuzefovich** [comment]
Sounds good, thanks.
